File: bindings/android/kotlin/org/ghananlp/speechid/GhanaSpeechId.kt
package org.ghananlp.speechid

import java.io.Closeable
import java.io.File

/**
 * Language identification over IPA phoneme strings.
 *
 * The phonemes come from sherpa-onnx running ghana-speech-phoneme-asr; this class takes it
 * from there. Pass the recogniser's output unmodified -- units are space separated and
 * multi-character units such as k͡p, kʰ and t͡ʃ must not be split.
 *
 * Not thread safe to create or close concurrently; [classify] and [posterior] are safe to
 * call from multiple threads on one instance.
 */
class GhanaSpeechId private constructor(private var handle: Long) : Closeable {

    /** Labels in model order. */
    val languages: List<String> =
        (0 until nativeNumLanguages(handle)).map { nativeLanguage(handle, it) ?: "?" }

    data class Result(
        val language: String,
        val index: Int,
        /**
         * Model score. Calibrated only for a logistic-regression head; with the SVM head
         * use [margin] instead as the confidence signal.
         */
        val confidence: Float,
        /** n-grams found in the vocabulary. */
        val matchedNGrams: Int,
    )

    /** Returns null when no n-gram matched -- there was no basis for a decision. */
    fun classify(ipa: String): Result? {
        val r = nativeClassify(handle, ipa.toByteArray(Charsets.UTF_8))
        val idx = r[0].toInt()
        if (idx < 0 || idx >= languages.size) return null
        return Result(languages[idx], idx, r[1], r[2].toInt())
    }

    /** Full posterior in model order, or null when nothing matched. */
    fun posterior(ipa: String): Map<String, Float>? {
        val p = nativeClassifyProbs(handle, ipa.toByteArray(Charsets.UTF_8)) ?: return null
        return languages.zip(p.toTypedArray()).toMap()
    }

    /** Top-1 minus top-2. A better confidence signal than the raw score for an SVM head. */
    fun margin(ipa: String): Float? {
        val p = nativeClassifyProbs(handle, ipa.toByteArray(Charsets.UTF_8)) ?: return null
        if (p.size < 2) return null
        p.sort()
        return p[p.size - 1] - p[p.size - 2]
    }

    override fun close() {
        if (handle != 0L) { nativeDestroy(handle); handle = 0L }
    }

    companion object {
        init { System.loadLibrary("ghana_speech_id_jni") }

        /**
         * @param modelDir directory holding head.onnx (or head.fp16.onnx), ngrams.txt,
         *   labels.txt and head_config.txt, as written by scripts/export_onnx.py.
         *   Assets must be extracted to real files first -- onnxruntime and the vocabulary
         *   reader both need filesystem paths, not an AssetManager stream.
         * @param threads onnxruntime intra-op threads; 1 is right for a phone.
         */
        @JvmStatic
        @JvmOverloads
        fun create(modelDir: File, threads: Int = 1): GhanaSpeechId {
            val onnx = File(modelDir, "head.onnx").let {
                if (it.exists()) it else File(modelDir, "head.fp16.onnx")
            }
            val ngrams = File(modelDir, "ngrams.txt")
            val labels = File(modelDir, "labels.txt")
            val config = File(modelDir, "head_config.txt")
            for (f in listOf(onnx, ngrams, labels)) {
                require(f.exists()) { "missing model file: ${f.name}" }
            }
            val err = arrayOfNulls<String>(1)
            val h = nativeCreate(
                onnx.absolutePath, ngrams.absolutePath, labels.absolutePath,
                if (config.exists()) config.absolutePath else "", threads, err,
            )
            check(h != 0L) { "could not create language identifier: ${err[0]}" }
            return GhanaSpeechId(h)
        }

        @JvmStatic private external fun nativeCreate(
            onnx: String, ngrams: String, labels: String, config: String,
            threads: Int, err: Array<String?>,
        ): Long
        @JvmStatic private external fun nativeDestroy(handle: Long)
        @JvmStatic private external fun nativeNumLanguages(handle: Long): Int
        @JvmStatic private external fun nativeLanguage(handle: Long, index: Int): String?
        @JvmStatic private external fun nativeClassify(handle: Long, ipa: ByteArray): FloatArray
        @JvmStatic private external fun nativeClassifyProbs(handle: Long, ipa: ByteArray): FloatArray?
    }
}

File: bindings/ios/GhanaSpeechID.swift
// Swift wrapper over the ghana-speech-id C API.
//
// The C header is imported directly through module.modulemap, so there is no Objective-C
// shim and no bridging header. Add libghana_speech_id.a and onnxruntime to the target,
// point the module map at include/ghana_speech_id.h, and this file is the whole binding.
//
//   let id = try LanguageIdentifier(modelDirectory: bundleURL)
//   if let r = id.classify(ipa: transcript) { print(r.language, r.confidence) }
//
// The IPA string comes from sherpa-onnx running ghana-speech-phoneme-asr; this library
// takes it from there. Units are space separated and multi-scalar units such as k͡p and t͡ʃ
// must stay whole, which they do as long as you pass the recogniser's output unmodified.

import Foundation

public struct LanguageResult: Sendable, Equatable {
    /// Predicted language label, e.g. "Twi_twi".
    public let language: String
    /// Index into `LanguageIdentifier.languages`.
    public let index: Int
    /// Model score for `language`. Calibrated only when the head was trained with
    /// logistic regression; for the SVM head prefer `margin` as a confidence signal.
    public let confidence: Float
    /// Number of n-grams that were found in the vocabulary.
    public let matchedNGrams: Int
}

public enum LanguageIdentifierError: Error, CustomStringConvertible {
    case missingArtefact(String)
    case createFailed(String)

    public var description: String {
        switch self {
        case .missingArtefact(let f): return "missing model file: \(f)"
        case .createFailed(let m):    return "could not create language identifier: \(m)"
        }
    }
}

public final class LanguageIdentifier {
    private let handle: OpaquePointer
    /// Labels in model order.
    public let languages: [String]

    /// - Parameters:
    ///   - modelDirectory: directory containing head.onnx (or head.fp16.onnx), ngrams.txt,
    ///     labels.txt and head_config.txt, as written by scripts/export_onnx.py.
    ///   - threads: onnxruntime intra-op threads. 1 is right for a phone; the head is a
    ///     single sparse gather and costs microseconds.
    public init(modelDirectory: URL, threads: Int32 = 1) throws {
        let fm = FileManager.default
        func path(_ name: String) -> String {
            modelDirectory.appendingPathComponent(name).path
        }
        var onnx = path("head.onnx")
        if !fm.fileExists(atPath: onnx) { onnx = path("head.fp16.onnx") }
        let ngrams = path("ngrams.txt")
        let labels = path("labels.txt")
        let config = path("head_config.txt")

        for f in [onnx, ngrams, labels] where !fm.fileExists(atPath: f) {
            throw LanguageIdentifierError.missingArtefact((f as NSString).lastPathComponent)
        }

        var cfg = GsidConfig()
        gsid_config_init(&cfg)
        cfg.num_threads = threads

        var err = [CChar](repeating: 0, count: 512)
        // The C strings must outlive gsid_create, so keep them alive for the whole call.
        let created: OpaquePointer? = onnx.withCString { pOnnx in
            ngrams.withCString { pNgrams in
                labels.withCString { pLabels in
                    config.withCString { pConfig in
                        cfg.onnx_path = pOnnx
                        cfg.ngrams_path = pNgrams
                        cfg.labels_path = pLabels
                        cfg.config_path = fm.fileExists(atPath: config) ? pConfig : nil
                        return gsid_create(&cfg, &err, err.count)
                    }
                }
            }
        }
        guard let h = created else {
            throw LanguageIdentifierError.createFailed(String(cString: err))
        }
        handle = h

        let n = Int(gsid_num_languages(h))
        languages = (0..<n).map { i in
            gsid_language(h, Int32(i)).map { String(cString: $0) } ?? "?"
        }
    }

    deinit { gsid_destroy(handle) }

    /// Returns nil when no n-gram matched, meaning there was no basis for a decision.
    /// Surface that as "unknown" rather than picking a language.
    public func classify(ipa: String) -> LanguageResult? {
        let r = ipa.withCString { gsid_classify(handle, $0) }
        guard r.index >= 0, Int(r.index) < languages.count else { return nil }
        return LanguageResult(language: languages[Int(r.index)],
                              index: Int(r.index),
                              confidence: r.confidence,
                              matchedNGrams: Int(r.num_matched))
    }

    /// Full posterior in model order, or nil when nothing matched.
    public func posterior(ipa: String) -> [String: Float]? {
        var buf = [Float](repeating: 0, count: languages.count)
        let n = ipa.withCString { s in
            buf.withUnsafeMutableBufferPointer { gsid_classify_probs(handle, s, $0.baseAddress) }
        }
        guard n > 0 else { return nil }
        return Dictionary(uniqueKeysWithValues: zip(languages, buf))
    }

    /// Top-1 score minus top-2. For the SVM head this separates confident from marginal
    /// decisions far better than the raw softmax value does.
    public func margin(ipa: String) -> Float? {
        var buf = [Float](repeating: 0, count: languages.count)
        let n = ipa.withCString { s in
            buf.withUnsafeMutableBufferPointer { gsid_classify_probs(handle, s, $0.baseAddress) }
        }
        guard n > 1 else { return nil }
        let sorted = buf.sorted(by: >)
        return sorted[0] - sorted[1]
    }

    public static var version: String { String(cString: gsid_version()) }
}
